Bottom: 64472f51a30be19a28c7a31471f070b430e2a556
Top:    fec4927466e87322393d52348fbe041073886997
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-09-05 15:17:46 +0800

vhost: avoid unnecessary iov iter operation

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index a608868..439e3ce 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -521,16 +521,18 @@ static void handle_tx(struct vhost_net *net)
 
 			/* Skip header. TODO: support TSO. */
 			len = iov_length(vq->iov, out);
-			iov_iter_init(&msg.msg_iter, WRITE, vq->iov, out, len);
-			iov_iter_advance(&msg.msg_iter, hdr_size);
-			/* Sanity check */
-			if (!msg_data_left(&msg)) {
-				vq_err(vq, "Unexpected header len for TX: "
-					"%zd expected %zd\n",
-					len, hdr_size);
-				goto out;
+			iov_iter_init(&msg.msg_iter, WRITE,  vq->iov, out, len);
+			if (unlikely(hdr_size)) {
+				iov_iter_advance(&msg.msg_iter, hdr_size);
+				/* Sanity check */
+				if (!msg_data_left(&msg)) {
+					vq_err(vq, "Unexpected header len for "
+						   "TX: %zd expected %zd\n",
+						len, hdr_size);
+					goto out;
+				}
+				len = msg_data_left(&msg);
 			}
-			len = msg_data_left(&msg);
 
 			zcopy_used = zcopy && len >= VHOST_GOODCOPY_LEN
 				     && (nvq->upend_idx + 1) % UIO_MAXIOV !=
